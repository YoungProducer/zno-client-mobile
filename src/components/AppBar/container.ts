/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 2 March 2020
 *
 * Declare main types and interfaces for AppBar component.
 * Create function which connect actions
 * and/or variables from the redux-store to the component.
 */

/** External imports */
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
import { green } from '@material-ui/core/colors';

/** Application's imports */

/** Define Material ui styles */
const styles = (theme: Theme) =>
    createStyles({
        root: {
            background: green[700],
            width: '100%',
            height: 50,
        },
        container: {
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            height: 'inherit',
        },
        navLink: {
            color: '#fff',
            fontSize: '1.2rem',
            marginRight: theme.spacing(3),
            '&:last-child': {
                marginRight: 0,
            },
        },
    });

/** Props which component get from the parent */
interface IOwnProps extends WithStyles<typeof styles>{}

/** Declare type which describe all props pushed to the component */
export type TAppBarProps =
    IOwnProps;

/** Export function to configure component via HOC */
export default compose(
    withStyles(styles, { withTheme: true }),
);
