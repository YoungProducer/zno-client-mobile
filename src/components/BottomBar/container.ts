/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 March 2020
 *
 * Container for BottomBar component.
 */

/** External imports */
import { connect } from 'react-redux';

/** Application's imports */
import { selectIsLoggedIn } from 'store/selectors/auth';
import { fetchLogoutAction } from 'store/actionCreators/auth';
import { RootState, setTestSuiteFinishedAction } from 'store/slices';

interface IStateProps {
    loggedIn: boolean;
}

interface IDispatchProps {
    logout: () => void;
    setTestSuiteFinished: (value: boolean) => void;
}

export type TBottomBarProps =
    IStateProps
    & IDispatchProps;

const mapStateToProps = (state: RootState): IStateProps => ({
    loggedIn: selectIsLoggedIn(state),
});

const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    logout: () => dispatch(fetchLogoutAction()),
    setTestSuiteFinished: (finished: boolean) =>
        dispatch(setTestSuiteFinishedAction(finished)),
});

export default connect<IStateProps, IDispatchProps>(
    mapStateToProps,
    mapDispatchToProps,
);
