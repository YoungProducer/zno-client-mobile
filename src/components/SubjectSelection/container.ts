/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Define main interfaces and types for SubjectSelection component.
 * Create function which connect actions
 * and/or variables from the redux0store to the component.
 */

/** External imports */
import { connect } from 'react-redux';

/** Application's imports */
import { fetchLogoutAction } from 'store/actionCreators/auth';
import { fetchSubjectsAction } from 'store/actionCreators/subjects';
import {
    selectSubjectsLoading,
    selectSubjectsList,
} from 'store/selectors/subjects';
import { selectIsLoggedIn } from 'store/selectors/auth';
import { RootState, TSubjectList } from 'store/slices';

/** Props which component get from the parent */
interface IOwnProps {}

/** Props which component select from the redux-store */
interface IStateProps {
    loading: boolean;
    subjectsList: TSubjectList;
    loggedIn: boolean;
}

/** Props(actions) which component can dispatch */
interface IDispatchProps {
    fetchSubjectsNames: () => void;
    fetchLogout: () => void;
}

/** Declare type which describe all props pushed to the component */
export type TSubjectSelectionProps =
    IOwnProps
    & IStateProps
    & IDispatchProps;

/** Create function which connect variables from the redux-store to component */
const mapStateProps = (state: RootState): IStateProps => ({
    loading: selectSubjectsLoading(state),
    subjectsList: selectSubjectsList(state),
    loggedIn: selectIsLoggedIn(state),
});

/** Create function which connect actions to the component */
const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    fetchSubjectsNames: () => dispatch(fetchSubjectsAction()),
    fetchLogout: () => dispatch(fetchLogoutAction()),
});

/**
 * Export function which connect actions
 * and/or variables from the redux-store to component.
 */
export default connect<IStateProps, IDispatchProps, IOwnProps>(
    mapStateProps,
    mapDispatchToProps,
);
