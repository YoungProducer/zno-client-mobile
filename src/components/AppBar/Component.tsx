/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 2 March 2020
 *
 * Header component which contain main routes and another information or actions.
 */

/** External imports */
import React from 'react';
import CssBaseLine from '@material-ui/core/CssBaseline';
import Container from '@material-ui/core/Container';
import AppBar from '@material-ui/core/AppBar';

/** Application's imports */
import { TAppBarProps } from './container';
import NavigationLink from 'components/NavigationLink';

/** Create component */
const Component = (props: TAppBarProps) => {
    const { classes } = props;

    return (
        <React.Fragment>
            <CssBaseLine />
            <AppBar
                position='static'
                className={classes.root}
            >
                <Container
                    maxWidth='lg'
                    className={classes.container}
                >
                    <NavigationLink
                        navLink={{
                            to: '/',
                        }}
                        className={classes.navLink}
                        data-testid='home-navlink'
                    >
                        Головна
                    </NavigationLink>
                    <NavigationLink
                        navLink={{
                            to: '/subject-selection',
                        }}
                        className={classes.navLink}
                        data-testid='tests-navlink'
                    >
                        Тести
                    </NavigationLink>
                </Container>
            </AppBar>
        </React.Fragment>
    );
};

export default Component;
