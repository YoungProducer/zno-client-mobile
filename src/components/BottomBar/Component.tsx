/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 March 2020
 *
 * Component which display main navigation of app.
 */

/** External imports */
import React, { useState, useMemo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import classNames from 'classnames';
import Collapse from '@material-ui/core/Collapse';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import ExpandLessIcon from '@material-ui/icons/ExpandLess';
import HomeIcon from '@material-ui/icons/Home';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import PersonIcon from '@material-ui/icons/Person';
import { makeStyles, createStyles, Theme } from '@material-ui/core';

/** Application's imports */
import { TBottomBarProps } from './container';

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            position: 'absolute',
            minHeight: 70,
            borderRadius: 9,
            background: '#fff',
            filter: 'drop-shadow(0px 3px 3.5px rgba(0,0,0,0.16))',
            bottom: theme.spacing(1),
            left: theme.spacing(1),
            right: theme.spacing(1),
            transform: `translateY(0)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        rootHidden: {
            transform: `translateY(60%)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        innerContainer: {
            width: '100%',
            height: '100%',
        },
        actionsContainer: {
            width: '100%',
            height: '100%',
            minHeight: 70,
            display: 'flex',
            justifyContent: 'space-around',
            alignItems: 'center',
        },
        expandIconButton: {
            zIndex: 10,
            position: 'absolute',
            left: '50%',
            transform: `translate(-50%, -60%)`,
            borderRadius: '50%',
            background: '#fff',
            filter: 'drop-shadow(0px 3px 3.5px rgba(0,0,0,0.05))',
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
            '&:hover': {
                background: '#fff',
            },
        },
        expandIconButtonHidden: {
            transform: `translate(-50%, -35%)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        expandIcon: {
            transform: `rotateX(180deg)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        expandIconHidden: {
            transform: `rotateX(0deg)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIconButton: {
            background: '#fff',
            boxShadow: `0px 3px 3.5px rgba(0,0,0,0)`,
            transition: theme.transitions.create(['box-shadow'], {
                duration: 300,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIconButtonActive: {
            boxShadow: `0px 3px 3.5px rgba(0,0,0,0.16)`,
            transition: theme.transitions.create(['box-shadow'], {
                duration: 500,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIcon: {
            color: '#69deac',
            width: 35,
            height: 35,
        },
        expandActions: {
            padding: theme.spacing(1),
            display: 'flex',
            justifyContent: 'space-between',
        },
        button: {
            color: '#fff',
            background: '#656cff',
            borderRadius: 4,
            width: '45%',
        },
    }));

const Component = ({ loggedIn, logout }: TBottomBarProps) => {
    const classes = useStyles({});

    const history = useHistory();
    const location = useLocation();

    const [hidden, setHidden] = useState<boolean>(false);
    const [expand, setExpand] = useState<boolean>(false);

    const toggleHidden = () => {
        if (expand) {
            setExpand(false);
            window.setTimeout(() => {
                setHidden(!hidden);
            }, 300);
        } else {
            setHidden(!hidden);
        }
    };
    const toggleExpand = () => setExpand(!expand);

    const handleSignIn = () => history.push('/auth/signin');
    const handleSignUp = () => history.push('/auth/signup');
    const handleHome = () => history.push('/');
    const handleProfile = () => history.push('/user/profile');

    const activeIcons = useMemo(() => {
        const { pathname } = location;

        return {
            home: pathname === '/' || pathname === '/subject-selection',
            profile: pathname === '/user/profile',
        };
    }, [location]);

    return (
        <div
            className={classNames(classes.root, {
                [classes.rootHidden]: hidden,
            })}
        >
            <IconButton
                className={classNames(classes.expandIconButton, {
                    [classes.expandIconButtonHidden]: hidden,
                })}
                onClick={toggleHidden}
            >
                <ExpandLessIcon className={classNames(classes.expandIcon, {
                    [classes.expandIconHidden]: hidden,
                })}/>
            </IconButton>
            <div className={classes.innerContainer}>
                <div className={classes.actionsContainer}>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: activeIcons.profile,
                        })}
                        onClick={handleProfile}
                    >
                        <PersonIcon className={classes.actionIcon}/>
                    </IconButton>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: activeIcons.home,
                        })}
                        onClick={handleHome}
                    >
                        <HomeIcon className={classes.actionIcon}/>
                    </IconButton>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: expand,
                        })}
                        onClick={toggleExpand}
                    >
                        <MoreHorizIcon className={classes.actionIcon}/>
                    </IconButton>
                </div>
                <Collapse in={expand} unmountOnExit>
                    <div className={classes.expandActions}>
                        { loggedIn && (
                            <Button
                                className={classes.button}
                                variant='contained'
                                disableElevation
                                onClick={logout}
                            >
                                Вийти
                            </Button>
                        )}
                        { !loggedIn && (
                            <>
                                <Button
                                    className={classes.button}
                                    variant='contained'
                                    disableElevation
                                    onClick={handleSignIn}
                                >
                                    Увійти
                                </Button>
                                <Button
                                    className={classes.button}
                                    variant='contained'
                                    disableElevation
                                    onClick={handleSignUp}
                                >
                                    Зареєструватися
                                </Button>
                            </>
                        )}
                    </div>
                </Collapse>
            </div>
        </div>
    );
};

export default Component;
