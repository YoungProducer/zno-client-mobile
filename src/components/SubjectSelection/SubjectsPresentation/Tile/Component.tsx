/**
 * Created by: Oleksandr Bezrukov
 * Creation datel: 24 March 2020
 *
 * Component which render task tiles with correct colors.
 */

/** External imports */
import React, { useState, useCallback, useEffect, useMemo } from 'react';
import { useHistory } from 'react-router-dom';
import classNames from 'classnames';
import Typography from '@material-ui/core/Typography';
import ButtonBase from '@material-ui/core/ButtonBase';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { yellow, red } from '@material-ui/core/colors';

/** Application's imports */
import { AdditionalAnswerPropertiesContext } from 'context/TestSuiteContext';
import { TTileProps } from './container';

export const tileWidth = 251;
export const activeTileWidth = 251;
/** Distance between tiles */
export const tilesSpacing = 16;

const useTileStyles = makeStyles((theme: Theme) =>
    createStyles({
        tile: {
            minWidth: tileWidth,
            width: tileWidth,
            height: 230,
            borderRadius: 17,
            overflow: 'overlay',
            boxShadow: `0px 0.5px 3.5px rgba(0, 0, 0, 0.16)`,
            background: '#edeff3',
            fontSize: '1.25rem',
            color: '#b19898',
            // filter: `drop-shadow(0px 3px 3.5px rgba(0,0,0,0.16))`,
            display: 'block',
            transition: theme.transitions.create(['height', 'width', 'background', 'color'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
            marginRight: tilesSpacing,
            '&:last-child': {
                marginRight: 0,
            },
        },
        tileActive: {
            // height: 230,
            width: activeTileWidth,
            minWidth: activeTileWidth,
            background: '#fff',
            filter: `drop-shadow(0px 3px 3.5px rgba(0,0,0,0.16))`,
            transition: theme.transitions.create(['height', 'width', 'background', 'color'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        tileHide: {
            visibility: 'hidden',
        },
        img: {
            width: '100%',
            height: 145,
            objectFit: 'contain',
            verticalAlign: 'top',
        },
        desc: {
            background: '#fff',
            height: 85,
            width: '100%',
            padding: theme.spacing(1),
        },
        label: {
            color: '#777b7e',
        },
    }));

const Tile = ({
    taskIndex,
    active,
    callback,
    hide,
    selected,
    gived,
    right,
    finished,
    image,
}: TTileProps) => {
    const classes = useTileStyles({});

    return (
        <div
            className={classes.tile}
            onClick={callback}
        >
            <img src={image} className={classes.img}/>
            <div className={classes.desc}>
                <Typography variant='h5' className={classes.label}>
                    Математика
                </Typography>
            </div>
        </div>
    );
};

export default Tile;
