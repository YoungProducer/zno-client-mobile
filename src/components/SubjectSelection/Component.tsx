/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Component to select subjects.
 */

/** External imports */
import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import classNames from 'classnames';
import AppBar from '@material-ui/core/AppBar';
import Container from '@material-ui/core/Container';
import Button from '@material-ui/core/Button';
import InputBase from '@material-ui/core/InputBase';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import SearchIcon from '@material-ui/icons/Search';
import { makeStyles, Theme, createStyles } from '@material-ui/core';

/** Application's imports */
import { TSubjectSelectionProps } from './container';
import SubjectTile from './SubjectTile';
import SubjectPresentation from './SubjectsPresentation';
import Logo from 'img/logo';

const useStyles = makeStyles((theme: Theme) => createStyles({
    root: {
        width: '100%',
        height: '100vh',
    },
    container: {
        display: 'flex',
        marginTop: theme.spacing(10),
    },
    searchBlock: {
        width: '50%',
    },
    popularBlock: {
        background: '#5c498c',
        height: 85,
        width: 'max-content',
        position: 'absolute',
        // bottom: '1%',
        bottom: '1%',
        right: 0,
        left: 0,
        margin: 'auto',
        borderRadius: 7,
        paddingLeft: theme.spacing(3),
        paddingRight: theme.spacing(3),
    },
    popularBlockInner: {
        height: 'inherit',
        width: '100%',
    },
    searchInput: {
        width: '100%',
        height: 67,
        background: '#4d8fcb',
        borderRadius: 34,
        paddingLeft: theme.spacing(2),
        paddingRight: 6,
        color: '#fff',
        fontSize: '1.3rem',
        marginTop: theme.spacing(8),
    },
    iconButton: {
        width: 55,
        height: 55,
        background: '#fff',
        borderRadius: '50%',
        '&:hover': {
            background: '#f1f1f1',
        },
    },
    icon: {
        color: '#5955cf',
        width: 30,
        height: 30,
    },
    typography: {
        color: '#fff',
    },
    actionsBlock: {
        display: 'flex',
        alignItems: 'center',
    },
    appBar: {
        background: 'none',
        paddingTop: theme.spacing(3),
        paddingBottom: theme.spacing(3),
    },
    appBarContainer: {
        display: 'flex',
        justifyContent: 'space-between',
    },
    appBarActions: {

    },
    signUpButton: {
        borderRadius: 25,
        background: '#fff',
        color: theme.palette.primary.main,
        paddingLeft: theme.spacing(2),
        paddingRight: theme.spacing(2),
        '&:hover': {
            background: '#f1f1f1',
        },
        minWidth: 100,
    },
    signInButton: {
        minWidth: 100,
        color: '#fff',
        '&:hover': {
            background: 'none',
        },
    },
}));

/** Create component */
const Component = (props: TSubjectSelectionProps) => {
    const classes = useStyles({});

    const history = useHistory();

    const {
        subjectsList,
        loggedIn,
        fetchSubjectsNames,
        fetchLogout,
    } = props;

    const [seachValue, setSearchValue] = useState<string>('');
    const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) =>
        setSearchValue(event.target.value);

    useEffect(() => {
        fetchSubjectsNames();
    }, []);

    return (
        <>
            <div className={classNames(classes.root, 'subject-selection-background')}>
                <AppBar elevation={0} className={classes.appBar} position='relative'>
                    <Container maxWidth='lg' className={classes.appBarContainer}>
                        <Logo />
                        <div className={classes.appBarContainer}>
                            { !loggedIn && (
                                <>
                                    <Button
                                        className={classes.signInButton}
                                        onClick={() => history.push('/auth/signin')}
                                    >
                                        Увійти
                                    </Button>
                                    <Button
                                        variant='text'
                                        className={classes.signUpButton}
                                        onClick={() => history.push('/auth/signup')}
                                    >
                                        Зареєструватися
                                    </Button>
                                </>
                            )}
                            { loggedIn && (
                                <Button
                                    variant='text'
                                    className={classes.signUpButton}
                                    onClick={fetchLogout}
                                >
                                    Вийти
                                </Button>
                            )}
                        </div>
                    </Container>
                </AppBar>
                <Container maxWidth='lg' className={classes.container}>
                    <div className={classes.actionsBlock}>
                        <div className={classes.searchBlock}>
                            <Typography
                                variant='h6'
                                className={classes.typography}
                            >
                                Для того щоб розпочати тест почніть вводити назву предемету у пошуку, після цього ви одразу побачите цей предмет
                                у правому блоці, просто натисніть на нього і ви будете направлені на сторінку тесту.
                            </Typography>
                            <InputBase
                                className={classes.searchInput}
                                placeholder='Знайти предмет'
                                value={seachValue}
                                onChange={handleSearchChange}
                                endAdornment={
                                    <IconButton
                                        size='small'
                                        className={classes.iconButton}
                                    >
                                        <SearchIcon
                                            className={classes.icon}
                                        />
                                    </IconButton>
                                }
                            />
                        </div>
                        <SubjectPresentation
                            subjectsList={subjectsList}
                            searchValue={seachValue}
                        />
                    </div>
                </Container>
            </div>
            {/* <div className={classes.popularBlock}>
                <SubjectTile
                    subject={subjectsList[0]}
                />
                <SubjectTile
                    subject={subjectsList[1]}
                />
            </div> */}
        </>
    );
};

export default Component;
