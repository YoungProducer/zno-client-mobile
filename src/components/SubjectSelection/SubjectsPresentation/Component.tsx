/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 23 March 2020
 *
 * Component which allows to select which task user want to complete.
 */

/** External imports */
import React, { useState, useCallback, useEffect, useMemo } from 'react';
import { useHistory } from 'react-router-dom';
import classNames from 'classnames';
import Box from '@material-ui/core/Box';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

/** Application's imports */
import Tile, {
    tileWidth,
    activeTileWidth,
    tilesSpacing,
} from './Tile';
import math from 'images/math.jpg';

const useComponentStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            overflow: 'hidden',
            position: 'relative',
            height: 300,
            width: '100%',
            cursor: 'pointer',
            marginLeft: theme.spacing(-1),
        },
        innerContainer: {
            display: 'flex',
            alignItems: 'flex-start',
            position: 'absolute',
            height: '100%',
            width: '100%',
            paddingLeft: theme.spacing(1),
        },
        animateContainer: {
            transition: theme.transitions.create('left', {
                duration: 500,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
    }));

const useCarouselFields = (props: ITaskSelectionProps) => {
    /** Destruct props */
    const {
        tasksAmount,
        activeTask,
        setTaskIndex,
    } = props;

    const history = useHistory();

    const handleTileClick = (id: string) => history.push(`/subject-configuration/${id}`);

    /** Offset of the inner container */
    const [offsetX, setOffestX] = useState<number>(0);
    const [prevStartX, setPrevStartX] = useState<number>(0);
    /** Toggles when mouse down or up */
    const [mouseDown, toggleMouseDown] = useState<boolean>(false);
    /** Related to animate class in inner container */
    const [animate, toggleAnimate] = useState<boolean>(false);

    const rootContainerRef = React.createRef<HTMLDivElement>();

    const [buttonsAmount, setButtonsAmount] = useState<number>(0);

    const minOffset = 0;
    const maxOffset = tasksAmount * (tileWidth + tilesSpacing) - tilesSpacing + (activeTileWidth - tileWidth);

    /**
     * If offset is out of range then
     * set offset in extreme positions.
     */
    const handleSetOffset = (offset: number) => {
        if (offset > minOffset) {
            setOffestX(minOffset);
        } else if (offset < -maxOffset + rootContainerRef.current.clientWidth) {
            setOffestX(-maxOffset + rootContainerRef.current.clientWidth);
        } else {
            setOffestX(offset);
        }
    };

    /**
     * Calculate diapason in which tiles must be displayed
     * to optimize carousel if tile index is out of range
     * to this tile will be added class with 'visibility: hidden'.
     */
    const getFirstTileIndex = useCallback(() => {
        const index = Math.floor(offsetX / (tileWidth + tilesSpacing)) * -1;
        return 0 > index ? 0 : index - 2;
    }, [offsetX]);

    const getLastTileIndex = useCallback(() => {
        const index = buttonsAmount + Math.floor(offsetX / (tileWidth + tilesSpacing)) * -1;
        return buttonsAmount > index ? buttonsAmount : index + 2;
    }, [buttonsAmount, offsetX]);

    /**
     * Handle mouse move to drag the carousel
     * offset divided by 2 to make carousel slower.
     */
    const handleOnTouchStart = (event: React.TouchEvent<HTMLDivElement>) => {
        setPrevStartX(event.touches[0].clientX);
    };

    const handleOnTouchMove = (event: React.TouchEvent<HTMLDivElement>) => {
        event.persist();
        const currentTouchX = event.touches[0].clientX;

        const mouseOffsetX = (prevStartX - currentTouchX) / -1;
        const newOffsetX = offsetX + mouseOffsetX;

        setPrevStartX(currentTouchX);

        handleSetOffset(newOffsetX);
    };

    /**
     * Toggle animate to 'false' to remove
     * animation class in innerContainer
     * to prevent bugs.
     */
    const handleOnMouseDown = () => {
        toggleMouseDown(true);
        toggleAnimate(false);
    };

    const handleOnMouseUp = () => toggleMouseDown(false);

    /**
     * If new active tile has border index
     * then toggle animate to 'true' to add
     * class in innerContainer which allows
     * to animate 'left' property and then
     * calculate and add offset to move
     * tiles to left or to right.
     */
    useEffect(() => {
        if (!mouseDown && activeTask !== 0) {
            /**
             * Difference between last tile and new active el.
             * 2 - lastTileIndex offset.
             */
            const lastDiff = getLastTileIndex() - 2 - activeTask;
            /**
             * Difference between first tile and new active el.
             * 2 - firstTileIndex offset.
             */
            const firstDiff = getFirstTileIndex() + 2 - activeTask;

            if (lastDiff < 3 && lastDiff >= 0) {
                toggleAnimate(true);
                handleSetOffset(Math.floor(offsetX - (buttonsAmount - 2) * (tileWidth + tilesSpacing)));
            }

            if (firstDiff < 3 && firstDiff >= 0) {
                toggleAnimate(true);
                handleSetOffset(Math.floor(offsetX + (buttonsAmount - 2) * (tileWidth + tilesSpacing)));
            }
        }
        if (activeTask <= getFirstTileIndex()) {
            const firstDiff = getFirstTileIndex() + 2;

            if (firstDiff !== 0) {
                toggleAnimate(true);
                handleSetOffset(Math.floor(offsetX + (getLastTileIndex() - 2) * (tileWidth + tilesSpacing)));
            }
        }
    }, [activeTask]);

    /**
     * Calculate buttons amount per innerContainer width.
     */
    useEffect(() => {
        const containerWidth = rootContainerRef.current.clientWidth;

        const amount = ((containerWidth + tilesSpacing) / (tileWidth + tilesSpacing));

        setButtonsAmount(Math.floor(amount));
    }, [rootContainerRef]);

    const tiles = useMemo(() =>
        [{ image: math, id: '5e69049424aa9a00072762eb' }].map((_, index) => (
            <Tile
                key={index}
                image={_.image}
                taskIndex={index}
                active={index === activeTask}
                hide={!(index >= getFirstTileIndex() && index <= getLastTileIndex())}
                callback={() => {
                    handleTileClick(_.id);
                }}
            />
        )), [
            activeTask,
            tasksAmount,
            getFirstTileIndex,
            getLastTileIndex,
        ]);

    return {
        handleOnTouchMove,
        handleOnMouseDown,
        handleOnMouseUp,
        handleOnTouchStart,
        tiles,
        animate,
        offsetX,
        rootContainerRef,
        buttonsAmount,
    };
};

export interface ITaskSelectionProps {
    tasksAmount: number;
    setTaskIndex: (index: number) => void;
    activeTask: number;
}

const Component = (props: ITaskSelectionProps) => {
    const classes = useComponentStyles({});

    const {
        offsetX,
        animate,
        tiles,
        rootContainerRef,
        handleOnTouchMove,
        handleOnTouchStart,
    } = useCarouselFields(props);

    return (
        <div
            className={classes.root}
            // onTouchStart={handleOnTouchStart}
            // onTouchMove={handleOnTouchMove}
            ref={rootContainerRef}
        >
            <Box
                className={classNames(classes.innerContainer, {
                    [classes.animateContainer]: animate,
                })}
                left={offsetX}
            >
                {tiles}
            </Box>
        </div>
    );
};

export default Component;
