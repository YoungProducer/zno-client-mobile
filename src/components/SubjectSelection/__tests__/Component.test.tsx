/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 3 March 2020
 *
 * Create test suites for SubjectSelection component.
 */

/** External imports */
import React from 'react';
import { shallow } from 'enzyme';

/** Application's imports */
import Component from '../Component';
import { TSubjectSelectionProps } from '../container';

/** Mock modules */
jest.mock('react-router-dom', () => ({
    useHistory: jest.fn(() => ({
        push: jest.fn(),
    })),
}));

describe('SubjectSelection component', () => {
    /** Create mock for function */
    const fetchSubjectsNames = jest.fn();
    const fetchLogout = jest.fn();

    /** Define required props */
    const requiredProps: TSubjectSelectionProps = {
        fetchSubjectsNames,
        fetchLogout,
        classes: {
            root: 'root',
        },
        loading: false,
        subjectsList: [{
            id: 'foo',
            name: 'math',
            image: 'bar',
        }],
        loggedIn: false,
    };

    afterEach(() => {
        (requiredProps.fetchSubjectsNames as jest.Mock).mockReset();
    });

    test('Is match snapshot', () => {
        /** Render component in shallow via enzyme */
        const tree = shallow(<Component {...requiredProps}/>);

        /** Assert component matches to snapshot */
        expect(tree).toMatchSnapshot();
    });
});
