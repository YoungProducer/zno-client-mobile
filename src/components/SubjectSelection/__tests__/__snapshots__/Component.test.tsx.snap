// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubjectSelection component Is match snapshot 1`] = `
<Fragment>
  <div
    className="makeStyles-root-1 subject-selection-background"
  >
    <WithStyles(ForwardRef(AppBar))
      className="makeStyles-appBar-11"
      elevation={0}
      position="relative"
    >
      <WithStyles(ForwardRef(Container))
        className="makeStyles-appBarContainer-12"
        maxWidth="lg"
      >
        <Component />
        <div
          className="makeStyles-appBarContainer-12"
        >
          <WithStyles(ForwardRef(Button))
            className="makeStyles-signInButton-15"
            onClick={[Function]}
          >
            Увійти
          </WithStyles(ForwardRef(Button))>
          <WithStyles(ForwardRef(Button))
            className="makeStyles-signUpButton-14"
            onClick={[Function]}
            variant="text"
          >
            Зареєструватися
          </WithStyles(ForwardRef(Button))>
        </div>
      </WithStyles(ForwardRef(Container))>
    </WithStyles(ForwardRef(AppBar))>
    <WithStyles(ForwardRef(Container))
      className="makeStyles-container-2"
      maxWidth="lg"
    >
      <div
        className="makeStyles-actionsBlock-10"
      >
        <div
          className="makeStyles-searchBlock-3"
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-typography-9"
            variant="h6"
          >
            Для того щоб розпочати тест почніть вводити назву предемету у пошуку, після цього ви одразу побачите цей предмет у правому блоці, просто натисніть на нього і ви будете направлені на сторінку тесту.
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(InputBase))
            className="makeStyles-searchInput-6"
            endAdornment={
              <WithStyles(ForwardRef(IconButton))
                className="makeStyles-iconButton-7"
                size="small"
              >
                <Memo(SearchIcon)
                  className="makeStyles-icon-8"
                />
              </WithStyles(ForwardRef(IconButton))>
            }
            onChange={[Function]}
            placeholder="Знайти предмет"
            value=""
          />
        </div>
        <Component
          searchValue=""
          subjectsList={
            Array [
              Object {
                "id": "foo",
                "image": "bar",
                "name": "math",
              },
            ]
          }
        />
      </div>
    </WithStyles(ForwardRef(Container))>
  </div>
</Fragment>
`;
